externals/kyuubi-trino-engine/src/main/scala/org/apache/kyuubi/engine/trino/schema/RowSet.scala

```diff
@@ -0,0 +1,320 @@
+        val tDoubleValue = new TDoubleValue
+        if (row(ordinal) != null) tDoubleValue.setValue(row(ordinal).asInstanceOf[Double])
+        TColumnValue.doubleVal(tDoubleValue)
+
+      case VARCHAR =>
+        val tStringValue = new TStringValue
+        if (row(ordinal) != null) tStringValue.setValue(row(ordinal).asInstanceOf[String])
+        TColumnValue.stringVal(tStringValue)
+
+      case _ =>
+        val tStrValue = new TStringValue
+        if (row(ordinal) != null) {
+          tStrValue.setValue(
+            toHiveString((row(ordinal), types(ordinal).getType), timeZone))
+        }
+        TColumnValue.stringVal(tStrValue)
+    }
+  }
+
+  /**
+   * A simpler impl of Trino's toHiveString
+   */
+  def toHiveString(dataWithType: (Any, String), timeZone: ZoneId): String = {
+    dataWithType match {
+      case (null, _) =>
+        // Only match nulls in nested type values
+        "null"
+
+      case (sd: java.sql.Date, DATE) =>
+        dateFormatter.format(sd.toLocalDate)
+
+      case (d: Date, DATE) =>
+        dateFormatter.format(d.toInstant)
+
+      case (ld: LocalDate, DATE) =>
+        dateFormatter.format(ld)
+
+      case (t: Timestamp, TIMESTAMP) =>
+        timestampFormatter.withZone(timeZone).format(t.toInstant)
+
+      case (i: Instant, TIMESTAMP) =>
+        timestampFormatter.withZone(timeZone).format(i)
+
+      case (t: Timestamp, TIMESTAMP_WITH_TIME_ZONE) =>
+        timestampWithZoneFormatter.withZone(timeZone).format(t.toInstant)
+
+      case (t: Time, TIME) =>
+        timeFormatter.format(t.toLocalTime)
+
+      case (lt: LocalTime, TIME) =>
+        timeFormatter.format(lt)
+
+      case (bin: Array[Byte], VARBINARY) =>
+        new String(bin, StandardCharsets.UTF_8)
+
+      case (decimal: java.math.BigDecimal, DECIMAL) =>
+        decimal.toPlainString
+
+      case (s: String, VARCHAR) =>
+        // Only match string in nested type values
+        "\"" + s + "\""
+
+      // for Array Map and Row, temporarily convert to string
+      // TODO further analysis of type
+      case (seq: Seq[_], _) =>
+        formatValue(seq)
+
+      case (m: Map[_, _], _) =>
+        formatValue(m)
+
+      case (row: Row, _) =>
+        formatValue(row)
+
+      case (other, _) =>
+        other.toString
+    }
+  }
+
+  def formatValue(o: Any): String = {
+    o match {
+      case null =>
+        "null"
+
+      case m: Map[_, _] =>
+        m.map { case (key, value) =>
+          formatValue(key) + ":" + formatValue(value)
+        }.toSeq.sorted.mkString("{", ",", "}")
+
+      case l: scala.collection.Seq[_] =>
+        l.map(formatValue).mkString("[", ",", "]")
+
+      case row: Row =>
+        row.getFields.asScala.map { r =>
+          val formattedValue = formatValue(r.getValue())
+          if (r.getName.isPresent) {
+            r.getName.get() + "=" + formattedValue
+          } else {
+            formattedValue
+          }
+        }.mkString("{", ",", "}")
+
+      case _ => o.toString
```

Comment:
what if we have a date/timestamp/decimal in complex types

Reply:
Can we take this as the next works? The string of type maybe complex and needs further analysis. We should call `tohivestring` recursively to analyze the type, `formatValue` is needless.